src/gluonts/dataset/parallelized_loader.py

```diff
@@ -0,0 +1,613 @@
+        start_index = int(
+            (worker_id / num_workers) * dataset_len
+        )  # calculate offsets for different replicas
+        end_index = (
+            None
+            if cyclic
+            else int(((worker_id + 1) / num_workers) * dataset_len)
+        )  # loop infinitely if cyclic
+        _worker_dataset.set_replica_info(
+            ReplicaInfo(
+                start_index=start_index,
+                end_index=end_index,
+                replica_id=worker_id,
+            )
+        )
+
+
+def sequential_sample_generator(dataset, transformation, is_train, cyclic):
+    while True:
+        for sample in transformation(data_it=dataset, is_train=is_train):
+            yield sample
+        # Dont cycle if not training time
+        if not cyclic:
+            return
+
+
+# TODO: for some reason cannot pickle 'future_observed_values' or 'future_target' when using InferenceDataLoader
+def _worker_fn(
+    batch_size: int,
+    batchify_fn: Callable,
+    dtype: DType,
+    is_train: bool,
+    cyclic: bool,
+    cycle_num: int,
+):
+    """Function for processing data in worker process."""
+    # pylint: disable=unused-argument
+    # it is required that each worker process has to fork a new MXIndexedRecordIO handle
+    # preserving dataset as global variable can save tons of overhead and is safe in new process
+
+    global _worker_dataset_iterator
+    global _worker_iterator_reset_num
+    global _worker_iterator_exhausted_indicator
+
+    # TODO: remove debug print
+    # print(
+    #     multiprocessing.current_process().name,
+    #     "iterator none:",
+    #     _worker_dataset_iterator is None,
+    # )
+
+    # initialize, or reset the iterator at each cycle
+    assert isinstance(_worker_iterator_reset_num, int)
+    if (_worker_iterator_reset_num < cycle_num) and (
+        cycle_num == 1 or not cyclic
+    ):
+        # TODO: remove debug print
+        # print(
+        #     "CYCLE NUM: ",
+        #     cycle_num,
+        #     "REPLICA ID:",
+        #     _worker_dataset.get_replica_info().replica_id,
+        #     "worker reset num: ",
+        #     _worker_iterator_reset_num,
+        #     "cycle_num",
+        #     cycle_num,
+        # )
+        _worker_reset_iterator(is_train, cyclic)
+
+    assert isinstance(
+        _worker_dataset_iterator, Iterable
+    ), "Dataset not Iterable."
+    transformed_data = list(
+        itertools.islice(_worker_dataset_iterator, batch_size)
+    )
+
+    if transformed_data:
+        success = True
+        batch = batchify_fn(
+            data=transformed_data, dtype=dtype, parallel_processing=True
+        )
+    else:
+        # the second time without being able to provide a batch we want to delay calling them again
+        # on fist exhaustion they should not be delayed, since they need to indicate depletion
+        if _worker_iterator_exhausted_indicator:
+            time.sleep(0.05)
```

Comment:
If im not missing something I think we could bump the penalty significantly here, since after each Validation pass through there is a Training phase, so this shouldn't interfere with the separate TrainingData loader. Basically same case for InferenceDataLoader, after which there is no second Inference using the same loader.
Some values to test could be '1' or '5'.

The main idea here is that if a depleted worker is requested again, we should prevent it from being called again so many times. Due to the current implementation there is no way to explicitly tell which worker should fetch a new batch.

Additionally, we don't want to terminate the worker, since for example the same ValidationDataLoader is called at every epoch. If we terminate the process one its depleted we would have to set up a new pool at each Validation iteration, which could come with costly copying etc.

Reply:
On the other hand, epochs can be very fast, so maybe .1-.2?

Reply:
It only needs to be so long that potential differences in processing time of different subsets of data are evened out a bit.